src/prettyPrinter.ts

```diff
@@ -1,804 +1,889 @@
-import JSONbig from "json-bigint";
+import * as A from "./grammar/ast";
+import { groupBy, intercalate, isUndefined } from "./utils/array";
+import { makeVisitor } from "./utils/tricks";
+
+//
+// Types
+//
+
+export const ppAstTypeId = A.idText;
+
+export const ppAstTypeIdWithStorage = (
+    type: A.AstTypeId,
+    storageType: A.AstId | null,
+): string => {
+    const alias = storageType ? ` as ${ppAstId(storageType)}` : "";
+    return `${ppAstTypeId(type)}${alias}`;
+};
+
+export const ppAstMapType = ({
+    keyType,
+    keyStorageType,
+    valueType,
+    valueStorageType,
+}: A.AstMapType): string => {
+    const key = ppAstTypeIdWithStorage(keyType, keyStorageType);
+    const value = ppAstTypeIdWithStorage(valueType, valueStorageType);
+    return `map<${key}, ${value}>`;
+};
+
+export const ppAstBouncedMessageType = ({
+    messageType,
+}: A.AstBouncedMessageType): string => `bounced<${ppAstTypeId(messageType)}>`;
+
+export const ppAstOptionalType = ({ typeArg }: A.AstOptionalType): string =>
+    `${ppAstType(typeArg)}?`;
+
+export const ppAstType = makeVisitor<A.AstType>()({
+    type_id: ppAstTypeId,
+    map_type: ppAstMapType,
+    bounced_message_type: ppAstBouncedMessageType,
+    optional_type: ppAstOptionalType,
+});
+
+//
+// Expressions
+//
+
+export const unaryOperatorType: Record<A.AstUnaryOperation, "post" | "pre"> = {
+    "+": "pre",
+    "-": "pre",
+    "!": "pre",
+    "~": "pre",
+
+    "!!": "post",
+};
+
+export const checkPostfix = (operator: A.AstUnaryOperation) =>
+    unaryOperatorType[operator] === "post";
 
 /**
- * Provides methods to format and indent Tact code.
+ * Description of precedence of certain type of AST node
  */
-export class PrettyPrinter {
+export type Precedence = {
     /**
-     * @param indentLevel Initial level of indentation.
-     * @param indentSpaces Number of spaces per indentation level.
+     * Add parentheses around `code` if in this `parent` position we need brackets
+     * @param check Position-checking function from parent
+     * @param code Code to put parentheses around
+     * @returns
      */
-    constructor(
-        private indentLevel: number = 0,
-        private readonly indentSpaces: number = 4,
-    ) {}
+    brace: (
+        position: (childPrecedence: number) => boolean,
+        code: string,
+    ) => string;
+    /**
+     * Create a position-checking function for positions where grammar rule recurses
+     * For example in `foo = "+" foo | bar` this should be used for `foo`
+     * In case of `"+" "+" bar` input, this won't put parentheses around inner
+     * `"+" bar`
+     */
+    self: (childPrecedence: number) => boolean;
+    /**
+     * Create a position-checking function for positions where grammar rule uses
+     * a higher-priority grammar rule
+     * For example in `foo = foo "-" bar | bar` this should be used for `bar`
+     * This enforces parentheses in expressions such as `x - (y - z)` where
```

Comment:
Well, precedence and associativity are cases of a more general thing: the grammar itself.

Rules can reference other rules or self-reference, and that's precisely what these methods are for.

Reply:
> what if we introduce the exponentiation operator

It will have a precedence of 105. In order to make it right-associative we'll have to swap `self` and `child` passed to `ppAstExpressionNested` of its arguments.

I'll try to improve wording of these comments, they're indeed confusing.